x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationMetrics.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.core.Strings;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.telemetry.metric.LongCounter;
+import org.elasticsearch.telemetry.metric.LongHistogram;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationResult;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+
+/**
+ * Provides a common way of registering and collecting various authentication metrics.
+ *
+ * @param <C> The type of context object which is used to attach additional attributes to collected metrics.
+ */
+final class AuthenticationMetrics<C> {
+
+    public static final String METRIC_SUCCESS_COUNT = "es.security.authc.%s.success.count";
+    public static final String METRIC_FAILURES_COUNT = "es.security.authc.%s.failures.count";
+    public static final String METRIC_AUTHENTICATION_TIME = "es.security.authc.%s.time";
```

Comment:
@pgomulka Since your review, I've decided to additionally introduced a new metric with name `es.security.authc.api_key.time`. The goal is to use it to measure individual authentication times (in nanoseconds). For measuring I'm using `LongHistogram`. I'd appreciate if you could have another look at this change. Thanks.

Reply:
not sure you can do this with any metrics to be fair. all of the metrics (even counters) are 'snapshoted' at meter event time (every meter interval). so you will not be able to see the per request authentication time - to for instance find out what user was causing long authenticatio ntime.
unless you want to drill down into authentication times per cluster/node?